XYZDelaunayMeshGenerator

roystgnr opened the pull request:
Pinging @eshemon ... and feeling embarrassed that I'm not remembering your team's Github ids; would you mind adding them yourself?

## Reason

This should close #28297 when it's ready to merge.  We don't yet have as much control in 3D as we did in 2D, but the new mesh generator here has at least all the basic functionality and the right place to hook in improvements

## Design

The new XYZDelaunayMeshGenerator uses the new libMesh Netgen shim to generate tetrahedral meshes in 3D the same way XYDelaunay did for triangular meshes in 2D.

## Impact

A new MeshGenerator option.

I'm going to mark this as draft for now; it deserves some user testing and developer review before we hit the merge button.  In particular, there are surely places where we're not giving our own error messages for bad input, we're just passing it along to Netgen and hitting their error messages, which may or may not be clear enough to users.

roystgnr committed:
XYZDelaunayGenerator mesh generator

Refs #28297 - this doesn't have as much functionality as XYDelaunay yet,
but it's at least got the basics working.

roystgnr committed:
Add XYZDelaunay regression tests

These might not be robust enough on a wide variety of systems; if that
happens we'll replace the exodiff tests with csvdiffs for invariants
like mesh volume.

roystgnr committed:
Doc page for XYZDelaunay

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2365112/) on 15bca74 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/28298/site)

This comment will be updated on new commits.

roystgnr commented:
And clang alone is enough to give us exodiffs.  I was afraid the reproduceability problems here might end up being even worse than in 2D.  I'll replace the exodiff tests with something weaker.

miaoyinb commented:
I will take a look. Thanks @roystgnr @eshemon

roystgnr committed:
Use CSVDiff not Exodiff in XYZDelaunay tests

We should probably add some additional moments here to avoid false
negatives, but at least this should be robust against false positives
from different Netgen behavior on different systems.

roystgnr committed:
Reference tetrahedralization issue in XYZ tests

miaoyinb commented on framework/src/meshgenerators/XYZDelaunayGenerator.C:
```diff
@@ -0,0 +1,405 @@
+{
+#ifdef LIBMESH_HAVE_NETGEN
+  // Put the boundary mesh in a local pointer
+  std::unique_ptr<UnstructuredMesh> mesh =
+      dynamic_pointer_cast<UnstructuredMesh>(std::move(_bdy_ptr));
```
I did some simples tests. The 3D meshes generated look great.
The mesh needs to be "prepared" here, otherwise the surface cannot be identified correctly.

Does this MG support 2D surface mesh (TRI3 elems) as input. I tried it and got an error.

Will do more tests and a more systematic review later this week.  Thanks.

roystgnr replied:
Thanks!

Yes, it needs neighbor pointers to be properly assigned ... I'll add a check for `!is_prepared()` and fix that up (probably at the libMesh level) when we find it's necessary.

What error did you get?  I've got a couple libMesh unit tests with Tri3 meshes as input.  They do also need neighbor pointers to have been assigned, *and* they need to be a closed manifold (every Tri3 needs to have neighbors on all sides).

miaoyinb replied:
I double checked my 2D input mesh and found a bug. I fixed it and now it works perfectly. Thanks.

roystgnr replied:
I'd still like to know what error message you got.  If it pointed to the underlying bug then we're in great shape, if it pointed to some consequence of the underlying bug then we might be able to improve that, and if it pointed to something completely unrelated then we ought to fix that.

miaoyinb replied:
I got a seg fault because the input mesh has extra TRI3 elements that do not belong to the surface.

roystgnr replied:
I get that you're the sort of user who can cope with my longstanding (bad) design philosophy of "As long as there's an assertion failure reporting a source file and line number then that should be all anyone needs to fix things, if not then that's a user Skill Issue" ... but MOOSE tries to be as friendly as possible to user errors even for less advanced users, and I'm trying to do better too.  :sweat_smile:

roystgnr replied:
A segfault is definitely unacceptable.  I'll see if I can replicate the problem.

roystgnr replied:
This should be resolved by https://github.com/libMesh/libmesh/pull/3919

roystgnr replied:
No, wait, the broken-boundary segfault was fixed, but the failure to handle unprepared meshes was not.  I'll get that now.

roystgnr replied:
Got that in libMesh/libmesh#3930

roystgnr committed:
Fix typo in XYZDelaunay docs

roystgnr commented:
That DistributedMesh breakage is going to need to be fixed at the libMesh level.

roystgnr commented:
A better error message and the distributed-mesh fixes are in libMesh, so I'll just get a new submodule going soon (once they get through CI to master there) and I won't add redundant fixes at this level.

miaoyinb commented:
It seems if we use volume elements for hole meshes, they have to be TET4? The following input leads to ` SYSTEM ERROR: more elements on face` if we skip the `[tet1]` block.

```
[Mesh]
  final_generator = triang
  [ccmg]
    type = ConcentricCircleMeshGenerator
    num_sectors = 8
    radii = '1'
    rings = '5'
    has_outer_square = off
    portion = left_half
    preserve_volumes = on
    smoothing_max_it = 3
  []
  [rg]
    type = RevolveGenerator
    input = ccmg
    axis_point = '0.0 0.0 0.0'
    axis_direction = '0.0 1.0 0.0'
    nums_azimuthal_intervals = 8
    revolving_angles = 180
  []
  [rg2]
    type = RevolveGenerator
    input = ccmg
    axis_point = '0.0 0.0 0.0'
    axis_direction = '0.0 1.0 0.0'
    nums_azimuthal_intervals = 8
  []
  [tet1]
    type = ElementsToTetrahedronsConverter
    input = rg2
  []
  [scale1]
    type = TransformGenerator
    input = tet1
    transform = SCALE
    vector_value = '0.1 0.1 0.1'
  []
  [t1]
    type = TransformGenerator
    input = scale1
    transform = TRANSLATE
    vector_value = '0.0 0.0 0.3'
  []
  [triang]
    type = XYZDelaunayGenerator
    boundary = 'rg'
    holes = 't1'
    stitch_holes = 'true'
    desired_volume = 1
  []
[]
```

roystgnr commented:
In theory we should be fine with non-tet holes.

In practice ... I'm not even making it to Netgen with that input - in the `volume_to_surface_mesh` conversion of the `rg` mesh, I get a `libmesh_assert_valid_unique_ids` error when a duplicate `node->unique_id()` is detected.  Have you run this in `dbg` mode yourself?

roystgnr commented:
Okay, I was expecting an issue in RevolveGenerator.C, since unique id handling for extrusion-type operations is tricky, but everything looks fine there.  This has to be my bug somehow.

roystgnr commented:
Possibly my new bug, if the parallel fixes that I haven't yet gotten into the MOOSE libMesh submodule are at fault...

roystgnr commented:
Aha!  No, it's a RevolveGenerator.C bug.

When we set a `unique_id()` before adding an object to a mesh, the mesh uses that id to update it's cache of the max `unique_id()` (on that processor, if it's distributed), so that `parallel_max_unique_id()` can be O(1) instead of O(N).

When we set a `unique_id()` *after* adding an object to a mesh, the mesh has no idea anything happened.  `parallel_max_unique_id()` is wrong.  Any code that uses it (including XYZDelaunay, and ironically RevolveGenerator) afterward is wrong.

There's a comment saying that RevolveGenerator was "Inspired by AdvancedExtruderGenerator", which used to have the same bug.  The simplest fix is just a `set_next_unique_id` call that takes into account the generator's formula for setting new unique ids.

miaoyinb commented:
Thanks for looking inti this. Let me see if I can fix the RevolveGenerator.

miaoyinb commented:
> Ping me as soon as you do? That's almost certainly _not_ the bug you're running into here, but it is interfering with me investigating that bug.

Is there a quick way to check/test if the `set_next_unique_id()` is done correctly?

miaoyinb commented:
Also, you can also try this input.It uses AdvancedExtruder instead. If you skip the `[tet1]`, it runs but the output mesh does not have that hole.

```
[Mesh]
    final_generator = triang
    [ccmg]
      type = ConcentricCircleMeshGenerator
      num_sectors = 8
      radii = '1'
      rings = '5'
      has_outer_square = off
      preserve_volumes = on
      smoothing_max_it = 3
    []
    [aeg]
      type = AdvancedExtruderGenerator
      input = ccmg
      heights = '1'
      num_layers = '10'
      direction = '0 0 1'
    []
    [aeg2]
      type = AdvancedExtruderGenerator
      input = ccmg
      heights = '1'
      num_layers = '5'
      direction = '0 0 1'
    []
    [tet1]
      type = ElementsToTetrahedronsConverter
      input = aeg2
    []
    [scale1]
      type = TransformGenerator
      input = tet1
      transform = SCALE
      vector_value = '0.2 0.2 0.2'
    []
    [t1]
      type = TransformGenerator
      input = scale1
      transform = TRANSLATE
      vector_value = '0.0 0.0 0.3'
    []
    [triang]
      type = XYZDelaunayGenerator
      boundary = 'aeg'
      holes = 't1'
      # stitch_holes = 'true'
      desired_volume = 1
    []
  []
```

roystgnr commented:
> Is there a quick way to check/test if the set_next_unique_id() is done correctly?

There really should be, shouldn't there?  I'll have to add a new test at the libMesh level at some point, to get run in `prepare_for_use()` maybe.

For right now, something like

```
for (auto elem : mesh.element_ptr_range())
  libmesh_assert(elem->unique_id() < mesh.next_unique_id());
for (auto node : mesh.node_ptr_range())
  libmesh_assert(node->unique_id() < mesh.next_unique_id());
```

Where by "like" I mean use `->` when `mesh` is a pointer, maybe fix whichever method names I've misspelled...

roystgnr commented:
> If you skip the [tet1], it runs but the output mesh does not have that hole.

Whether or not I skip the `[tet1]` I get a `NetGen failed to generate any tetrahedra` error, unless I comment out the `holes = 't1'`.

And ... I can't blame it?  If I generate `triang` with no holes then I see bounding box extents of {(0.48461,0.616791,0.8),(0.709385,0.834149,0.9)}, whereas if I generate `t1` I see {(-0.200644,-0.200644,0.3),(0.200646,0.200644,0.5)}.  That's not a hole, that's not even intersecting.

Should I add a bounding box sanity check here?  I can probably do it inside an existing element loop to avoid it being too expensive.  It would have false negatives (it's possible for a hole's bounding box to be contained within an outer boundary's bounding box despite the hole not being contained in the outer boundary) but it would at least catch some of these cases with a more informative error message.

miaoyinb commented:
Ah, I think you missed the preparation as we discussed before. 

```
  if (!mesh->is_prepared())
    mesh->prepare_for_use();
```

In fact, I added `prepare_for_use` for holes and it works now even when `[tet1]` is skipped.

miaoyinb commented:
And now it also works for the revolving case.

roystgnr commented:
Gah!  We did talk about that and I didn't get it in to libMesh yet.  Well, the bright side is that I already wrote the bounding box sanity check that you shouldn't have failed, so now we'll have both the preparedness fix *and* the bounding box sanity check in my next PR there.  :-P

roystgnr commented:
The "DistributedMesh" bug appears to be a Netgen bug.  All DistributedMesh is doing to trigger it is repartitioning in many different ways while changing the mesh numbering to match, which changes the order in which we hand boundary elements and points to Netgen.  This is much more serious than an actual distributed-mesh bug would have been; I can now replicate it in a unit test in serial...

roystgnr committed:
Add test for stitching failures to XYZDelaunay

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2445307/) on 5d4c518 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>b9e8bd</th>
      <th colspan=4>#28298 5d4c51</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.05%</td>
      <td>85.05%</td>
      <td>+0.00%</td>
      <td>85.08%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>105317</td>
      <td>105471</td>
      <td>+154</td>
      <td>154</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>18509</td>
      <td>18536</td>
      <td>+27</td>
      <td>27</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/28298/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/28298/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/28298/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/28298/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/28298/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/28298/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/28298/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/28298/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/28298/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/28298/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/28298/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/28298/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/28298/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/28298/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/28298/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/28298/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/28298/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/28298/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/28298/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/28298/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/28298/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/28298/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/28298/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/28298/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/28298/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/28298/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/28298/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/28298/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/28298/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/28298/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 85.08% is less than the suggested 90.0%



This comment will be updated on new commits.

GiudGiud commented on framework/doc/content/source/meshgenerators/XYZDelaunayGenerator.md:
```diff
@@ -0,0 +1,42 @@
+An input mesh will be used to define the outer boundary of a
+generated tetrahedral output mesh.  Additional interior vertices can
+be automatically generated by specifying a desired tetrahedral volume.
+Additional interior boundary "holes" can be meshed by the mesh
```
meshed or excluded ?

GiudGiud commented on framework/doc/content/source/meshgenerators/XYZDelaunayGenerator.md:
```diff
@@ -0,0 +1,42 @@
+to a given uniform element volume specified as
+[!param](/Mesh/XYZDelaunayGenerator/desired_volume).
+This refinement is currently controlled by the third-party Netgen
+code.
```
```suggestion
code.

!syntax parameters /Mesh/XYZDelaunayGenerator

!syntax inputs /Mesh/XYZDelaunayGenerator


```

GiudGiud commented on test/tests/meshgenerators/xyz_delaunay_generator/tests:
```diff
@@ -0,0 +1,55 @@
+      expect_err = "The hole meshes contain subdomain name maps with conflicts."
+      detail = "if the input hole meshes contain subdomain name maps with conflicts"
+    []
+  []
```
should it support an open domain? With a boundary that is not closed

roystgnr replied:
No, and it's got a check on the libMesh side for that.  Looking at that again, though ... any suggestions for improvements on the error message?

```
       "Missing element when trying to identify boundary component to mesh.\n"
       "Perhaps a boundary element does not have all neighbor pointers set?\n"
```

That "does not have all neighbor pointers set?" suggestion is probably more of a red herring at this point, since I stopped assuming that input meshes were prepared and started fixing up neighbor pointers myself.

I'm thinking something like:

```
       "Tet generation encountered a 2D element with a null neighbor, but a\n"
       "boundary surface must be a closed 2D manifold.\n"
```

GiudGiud replied:
is manifold too mathy for users?
could do 2D closed manifold (surface) ?

"with a null neighbor outside of the prescribed holes and boundaries"

roystgnr replied:
I'll add "(surface)".

"outside of the prescribed holes and boundaries" isn't actually correct, though - both the outer boundary and each hole boundary should not have any boundaries.  ∂∘∂=0 for the boundary operator on manifolds-with-boundary, in mathy parlance.  ;-)

GiudGiud commented on framework/src/meshgenerators/XYZDelaunayGenerator.C:
```diff
@@ -0,0 +1,408 @@
+      algorithm,
+      "Control the use of binary search for the nodes of the stitched surfaces.");
+  params.addParam<bool>(
+      "verbose_stitching", false, "Whether mesh stitching should have verbose output.");
```
```suggestion
      "verbose_stitching", false, "Whether mesh hole stitching should have verbose output.");
```

GiudGiud commented on framework/src/meshgenerators/XYZDelaunayGenerator.C:
```diff
@@ -0,0 +1,408 @@
+    // How did we never add a ReplicatedMesh(MeshBase&) constructor in
+    // libMesh?
+    const UnstructuredMesh & hole = dynamic_cast<UnstructuredMesh &>(**hole_ptr);
+    ngholes->push_back(std::make_unique<ReplicatedMesh>(hole));
```
will this serialize/replicate or should we be checking they are replicated already?

roystgnr replied:
The `ReplicatedMesh(UnstructuredMesh&)` constructor handles it ... as of https://github.com/libMesh/libmesh/pull/3921 anyway.

GiudGiud commented on framework/src/meshgenerators/XYZDelaunayGenerator.C:
```diff
@@ -0,0 +1,408 @@
+      }
+    }
+
+  // Assign new subdomain name, if provided
+  if (isParamValid("output_subdomain_name"))
+    mesh->subdomain_name(_output_subdomain_id) = getParam<SubdomainName>("output_subdomain_name");
```
```suggestion
```
already above

GiudGiud commented on framework/src/meshgenerators/XYZDelaunayGenerator.C:
```diff
@@ -0,0 +1,408 @@
+{
+  if (!_stitch_holes.empty() && _stitch_holes.size() != _hole_ptrs.size())
+    paramError("stitch_holes", "Need one stitch_holes entry per hole, if specified.");
+}
```
should we check hole_boundaries.size() = _hole_ptrs.size()

GiudGiud commented on framework/src/meshgenerators/XYZDelaunayGenerator.C:
```diff
@@ -0,0 +1,408 @@
+    auto hole_boundary_ids = MooseMeshUtils::getBoundaryIDs(*mesh, hole_boundaries, true);
+
+    if (hole_boundary_ids.size() != _hole_ptrs.size())
+      paramError("hole_boundary_ids", "Need one hole_boundary_ids entry per hole, if specified.");
```
this parameter doesnt exist. I m thinking check is better in constructor too?

GiudGiud commented on framework/src/meshgenerators/XYZDelaunayGenerator.C:
```diff
@@ -0,0 +1,408 @@
+  for (auto & hole_ptr : _hole_ptrs)
+    hole_ptr->reset();
+
+  mesh->prepare_for_use();
```
I think best practice now is to set_isnt_prepare and prepare only when needed in the MGs that require it

roystgnr committed:
Remove redundant subdomain_name setter

roystgnr committed:
Fix paramError parameter, move to constructor

roystgnr committed:
Use `set_isnt_prepared()` after generation

roystgnr committed:
Apply suggestions from code review

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

roystgnr committed:
holes can be "meshed"->"specified" in docs

roystgnr commented:
@GiudGiud , I think that addresses everything, except for the change in error message on the libMesh side.

GiudGiud commented:
Added some parallel sweep tests for peace of mind
